### PR TITLE
settings: Make page scrollable

### LINKED ISSUE #1904: Settings page should scroll
The settings page should be scrollable if the content doesn't all fit on the screen.

Currently there are only a few settings, so for most users this doesn't happen. But it can happen if the user has a very small screen (https://github.com/zulip/zulip-flutter/issues/1898#issuecomment-3386032763) or have their display zoomed in for accessibility. In the future it'll probably happen for more users, as we add more settings.

### PR DESCRIPTION
**Pull request by MritunjayTiwari14**
| Before | After (Screenshot) |
| ------- | ------------------ |
| column_l | listview_l |
| column_d | listview_d |

<br>

### Before

https://github.com/user-attachments/assets/03716367-b5d9-4f83-84f4-3716abb59829

<br>

### After (Normal)

https://github.com/user-attachments/assets/92a05eab-6789-4811-bb95-271838c864f3

<br>

### After (Increased Text Size to unreasonably high in android settings to test edge case)

https://github.com/user-attachments/assets/c070507f-8a4f-4fcf-8f4f-5731bef5f9d3

Fixes #1904

### CODE REVIEW

**Comment by gnprice** [comment]
Thanks for taking care of this!

Before we can review this change, it will need a test.

Those screenshots and recordings are very helpful, thanks. One remark: there's nothing unreasonable about the large text size — people vary in how sharp their vision is, and for some people that setting is useful by making it possible to read the text on the screen.

**Comment by MritunjayTiwari14** [comment]
Thank you @gnprice for reminding about the test.  I have added the test for scrolling in the latest commit.

I have limited the screen size to 200x200px to test the scrolling effect as lower down to 100x100px caused some rendering issue while running the test.

File: test/widgets/settings_test.dart [hunk 1]
```diff
@@ -160,6 +160,24 @@ void main() {
     }, variant: TargetPlatformVariant({TargetPlatform.android, TargetPlatform.iOS}));
   });
 
+  group('SettingPageScrollBehavior', () {
```
Comment (chrisbobbe) [inline test/widgets/settings_test.dart]: This string looks like an identifier, like the name of a class or something, but then I don't find anything when I look for its definition. To be less confusing, we should just say "SettingsPage scroll behavior" or similar. (`SettingsPage` is the name of a class, so it doesn't add confusion to use that name.)

Comment (chrisbobbe) [reply test/widgets/settings_test.dart]: Also nit: the test groups are now in this order:

1. 'ThemeSetting'
2. 'BrowserPreference'
3. _(this group)_
4. 'VisitFirstUnreadSetting'
5. 'MarkReadOnScrollSetting'

To be more organized, let's put the new group before the groups that are about individual settings, instead of in the middle of them.

File: test/widgets/settings_test.dart [hunk 2]
```diff
@@ -160,6 +160,24 @@ void main() {
...
+    testWidgets('scroll settings list when screen size is small', (tester) async {
+      tester.view.physicalSize = const Size(200, 200);
+      tester.view.devicePixelRatio = 1.0;
+      addTearDown(tester.view.reset);
+      await prepare(tester);
+
+      final lastElementFinder = GlobalSettingsStore.experimentalFeatureFlags.isNotEmpty
+          ? find.text("Experimental features")
+          : find.text("Mark messages as read on scroll");
+      check(lastElementFinder.evaluate().isEmpty).equals(true);
```
Comment (chrisbobbe) [inline test/widgets/settings_test.dart]: nit: how about `check(lastElementFinder).findsNothing();`

File: test/widgets/settings_test.dart [hunk 3]
```diff
@@ -160,6 +160,24 @@ void main() {
...
+
+      await tester.scrollUntilVisible(lastElementFinder, 100,
+        scrollable: find.byType(Scrollable));
+      check(lastElementFinder.evaluate().isEmpty).equals(false);
```
Comment (chrisbobbe) [inline test/widgets/settings_test.dart]: nit: how about `check(lastElementFinder).findsOne();

File: test/widgets/settings_test.dart [hunk 4]
```diff
@@ -160,6 +160,24 @@ void main() {
...
+  group('SettingPageScrollBehavior', () {
+    testWidgets('scroll settings list when screen size is small', (tester) async {
+      tester.view.physicalSize = const Size(200, 200);
+      tester.view.devicePixelRatio = 1.0;
+      addTearDown(tester.view.reset);
+      await prepare(tester);
+
+      final lastElementFinder = GlobalSettingsStore.experimentalFeatureFlags.isNotEmpty
+          ? find.text("Experimental features")
+          : find.text("Mark messages as read on scroll");
```
Comment (chrisbobbe) [inline test/widgets/settings_test.dart]:
```suggestion
      final lastElementFinder = GlobalSettingsStore.experimentalFeatureFlags.isNotEmpty
        ? find.text("Experimental features")
        : find.text("Mark messages as read on scroll");
```

File: test/widgets/settings_test.dart (hunk 2 above)
Comment (chrisbobbe) [reply test/widgets/settings_test.dart]: (Good thought to test that the row isn't visible before scrolling. We want to be sure that the scrolling is actually what causes the row to become visible, rather than some other factor, which could arise in the future, like the settings being presented in a different order than when we wrote this test.)

File: test/widgets/settings_test.dart [hunk 5]
```diff
@@ -160,6 +160,24 @@ void main() {
     }, variant: TargetPlatformVariant({TargetPlatform.android, TargetPlatform.iOS}));
   });
 
+  group('SettingPageScrollBehavior', () {
+    testWidgets('scroll settings list when screen size is small', (tester) async {
```
Comment (chrisbobbe) [inline test/widgets/settings_test.dart]: nit:

```suggestion
    testWidgets('content is scrollable when taller than a screenful', (tester) async {
```

(Helps focus the reader on what the test's purpose is, and acknowledges that the test can still be helpful when the screen is normal-sized, in a future where the list of settings is much longer.)

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Comments below.

For the commit message:

```
settings: Add scroll for settings page

This commit replaced Column with
ListView Widget to enable scrolling effect.
```

- It needs a `Fixes #<issue number>.` line, like other commits that fix issues in the tracker.
- The sentence in the body isn't needed; it just restates what's already obvious from reading the diff.
- For the summary line, "add scroll" sounds a little odd to me; how about `settings: Make page scrollable`.

File: test/widgets/settings_test.dart (hunk 1 above)
Comment (MritunjayTiwari14) [reply test/widgets/settings_test.dart]: Yes, Moving it to the top seems reasonable here.

File: test/widgets/settings_test.dart (hunk 3 above)
Comment (MritunjayTiwari14) [reply test/widgets/settings_test.dart]: Seems more high level.

**Comment by MritunjayTiwari14** [comment]
Thank you @chrisbobbe for taking time to review, all the requested changes and some other nits have been pushed.

**Comment by chrisbobbe** [comment]
Bump on https://github.com/zulip/zulip-flutter/pull/1911#pullrequestreview-3373037117 :
> * It needs a `Fixes #<issue number>.` line, like other commits that fix issues in the tracker.

Bump on https://github.com/zulip/zulip-flutter/pull/1911#discussion_r2457609946 :
> […] we should just say "SettingsPage scroll behavior" or similar. (`SettingsPage` is the name of a class, so it doesn't add confusion to use that name.)

**Comment by MritunjayTiwari14** [comment]
Thanks @chrisbobbe for pointing out the issue. The following has been corrected

- Corrected the Class name `SettingsPage`
- Added `Fixes #1904` in the commit description.

**Review by chrisbobbe — CHANGES_REQUESTED** [review]
Thanks!

From the Git history (see [Greg's "secret" to using `git log -p`](https://github.com/zulip/zulip-mobile/blob/main/docs/howto/git.md#2-the-secret-to-using-git-log--p), you can see we tend to use
1. "Fixes #1904." at the top, or
2. "Fixes: #1904" at the bottom

but not "Fixes #1904". (I think we're converging on option 1, which is what the Flutter project uses?)

**Comment by gnprice** [comment]
Yeah, I like option 1 of those. (That's what Flutter upstream uses, and I like how it sets context for what the rest of the commit message will be about.)

It's probably past time we should take the old style guide from zulip-mobile, copy the parts that apply here, and add that to this repo. This will be a good thing to write there.

**Comment by MritunjayTiwari14** [comment]
Thank you @chrisbobbe and @gnprice for the clarification, the commit description has been updated with the option 1 as recommended:

> settings: Make page scrollable
> Fixes #1904

**Comment by chrisbobbe** [comment]
"Fixes #1904."

**Comment by MritunjayTiwari14** [comment]
Thank you @chrisbobbe for noticing the typo. Commit description changed to:

> settings: Make page scrollable
> Fixes #1904.

**Comment by chrisbobbe** [comment]
Marking for Greg's review; thank you.

File: lib/widgets/settings.dart [hunk 6]
```diff
@@ -20,7 +20,7 @@ class SettingsPage extends StatelessWidget {
     return Scaffold(
       appBar: ZulipAppBar(
         title: Text(zulipLocalizations.settingsPageTitle)),
-      body: Column(children: [
+      body: ListView(children: [
```
Comment (gnprice) [inline lib/widgets/settings.dart:23]: settings: Make page scrollable
    Fixes #1904.

The body should be separated from the summary line by a blank line, so that it makes its own paragraph.

File: test/widgets/settings_test.dart [hunk 7]
```diff
@@ -64,6 +64,24 @@ void main() {
...
+      addTearDown(tester.view.reset);
+      await prepare(tester);
+
+      final lastElementFinder = GlobalSettingsStore.experimentalFeatureFlags.isNotEmpty
+        ? find.text("Experimental features")
+        : find.text("Mark messages as read on scroll");
+      check(lastElementFinder).findsNothing();
+
+      await tester.scrollUntilVisible(lastElementFinder, 100,
+        scrollable: find.byType(Scrollable));
```
Comment (gnprice) [inline test/widgets/settings_test.dart]: This value for the `scrollable` option is the default behavior, right? So can leave it out.

File: test/widgets/settings_test.dart [hunk 8]
```diff
@@ -64,6 +64,24 @@ void main() {
         hasCheckedState: true, isChecked: expectedIsChecked);
   }
 
+  group('SettingsPage scroll behavior', () {
+    testWidgets('content is scrollable when taller than a screenful', (tester) async {
```
Comment (gnprice) [inline test/widgets/settings_test.dart]: nit: can simplify by not wrapping this test in a group (since it was the only thing there anyway):
```suggestion
  testWidgets('SettingsPage content is scrollable when taller than a screenful', (tester) async {
```

**Review by gnprice — COMMENTED** [review]
Thanks @MritunjayTiwari14 for taking care of this, and thanks @chrisbobbe for the previous reviews! Just small comments below.

**Comment by MritunjayTiwari14** [comment]
Thank you @gnprice for the review. The following has been changed in the latest commit:

- Unwrapped the test `content is scrollable when taller than a screenful` out of group `SettingsPage scroll behavior` and removed the group.

- Added a new line between the summary and body so that `Fixes #1904.` appears at the top of body not along with summary.

- Removed `scrollable` from `tester.scrollUntilVisible` since flutter automatically finds the Scrollable by default for `scrollable`.

File: test/widgets/settings_test.dart (hunk 8 above)
Comment (gnprice) [reply test/widgets/settings_test.dart]: bump — I do think the "SettingsPage" part is helpful, for setting context when this test case goes by in `flutter test` output

**Comment by MritunjayTiwari14** [comment]
Yes @gnprice, it would be right to provide context instead of just saying content, even though the test belong to SettingsPage . The context of content within SettingsPage can vary. So in the latest changes the following has been accommodated:

- Changed `content` to `SettingsPage` to provide a better context of test.

**Comment by gnprice** [comment]
Thanks! Looks good; merging.
